### PR TITLE
5065 empty map with document reference

### LINKED ISSUE #5065: Document with Map decorated with @NonNull and @DocumentReference fails to deserialize
Given the following class:
```java
package org.example.mongoemptymaptest.model;

import lombok.Builder;
import lombok.Data;
import lombok.NonNull;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;
import org.springframework.data.mongodb.core.mapping.DocumentReference;

import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

@Data
@Builder
@Document(collection = "parent_documents")
public class ParentDocument {

    @Id
    @Builder.Default
    UUID id = UUID.randomUUID();

    @NonNull
    @Builder.Default
    @DBRef
    Map<String, ChildDocument> dbRefChildren = new HashMap<>();

    @NonNull
    @Builder.Default
    @DBRef(lazy = true)
    Map<String, ChildDocument> lazyDbRefChildren = new HashMap<>();

    @NonNull
    @Builder.Default
    @DocumentReference
    Map<String, ChildDocument> documentReferenceChildren = new HashMap<>();

    @NonNull
    @Builder.Default
    @DocumentReference(lazy = true)
    Map<String, ChildDocument> lazyDocumentReferenceChildren = new HashMap<>();
}
```

The document will fail to deserialize if `documentReferenceChildren` is an empty map, and will fail to properly resolve the map for `lazyDocumentReferenceChildren` if it is an empty map.  In that case, `getLazyDocumentReferenceChildren().values()` method will return null.

Unit tests illustrating the bug: https://github.com/bamapookie/MongoEmptyMapTest

### PR DESCRIPTION
**Pull request by bamapookie**
- [x] You have read the [Spring Data contribution guidelines](https://github.com/spring-projects/spring-data-build/blob/master/CONTRIBUTING.adoc).
- [x] You use the code formatters provided [here](https://github.com/spring-projects/spring-data-build/tree/master/etc/ide) and have them applied to your changes. Don’t submit any formatting related changes.
- [x] You submit test cases (unit or integration tests) that back your changes.
- [x] You added yourself as author in the headers of the classes you touched. Amend the date range in the Apache license header if needed. For new types, add the license header (copy from another file and set the current year only).

Copyright date already set to 2025.  No new files. No author lists in files I touched, so I left my name out.

Fixes #5065

### CODE REVIEW

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java [hunk 1]
```diff
@@ -478,4 +481,64 @@ public String toString() {
...
+		void readsEmptyMapWithDocumentReferenceCorrectly() {
+			org.bson.Document document = org.bson.Document.parse("{\"map\":{}}");
+			DocumentReferenceEmptyMapDocument target = converter.read(DocumentReferenceEmptyMapDocument.class, document);
+			assertThat(target.map).isNotNull().isEmpty();
+		}
+
+		@Test // GH-5065
+		@DisplayName("Converter should read an empty object as an empty map with a valis values property when using @DocumentReference(lazy = true).")
+		void readsEmptyMapWithLazyLoadedDocumentReferenceCorrectly() {
+			org.bson.Document document = org.bson.Document.parse("{\"map\":{}}");
```
Comment (christophstrobl) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java]: I think it would also be interesting to have test for `{ 'map' : null }` expecting `null` for the `target.map`.

**Review by christophstrobl — COMMENTED** [review]
Thank you for taking the time to prepare this PR. 
Kindly let me know if you'd like to add the tests from the review comment. If not, I can take it from here and continue the work.

**Comment by bamapookie** [comment]
I added lower level tests in the PR that didn't require the @NonNull
annotations to test.  They really only caused the issue to fail faster.  I
can add a test for `{ map: null }`

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java [hunk 2]
```diff
@@ -478,4 +481,130 @@ public String toString() {
 					+ ", localTime=" + this.getLocalTime() + ", localDateTime=" + this.getLocalDateTime() + ")";
 		}
 	}
+
+	@Nested // GH-5065
+	@DisplayName("GH-5065: Empty Map Tests")
```
Comment (schauder) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java]: No `@DisplayName` please.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java [hunk 3]
```diff
@@ -478,4 +481,130 @@ public String toString() {
...
+			assertThat(converter.read(LazyDocumentReferenceEmptyMapDocument.class, document).getMap()).isNotNull().isEmpty();
+			assertThat(converter.read(LazyDocumentReferenceEmptyMapDocument.class, document).map.values()).isNotNull().isEmpty();
+			assertThat(converter.read(LazyDocumentReferenceEmptyMapDocument.class, document).getMap().values()).isNotNull().isEmpty();
+		}
+
+		@Test // GH-5065
+		@DisplayName("Converter should read an empty object as an empty map when using @DBRef.")
+		void readsEmptyMapWithDBRefCorrectly() {
+			org.bson.Document document = org.bson.Document.parse("{\"map\":{}}");
+			//DBRefEmptyMapDocument target = converter.read(DBRefEmptyMapDocument.class, document);
```
Comment (schauder) [inline spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java]: What's up with the commented code line, and the repetition of that expression in every assert?

Why not introduce a `result` variable and assert on that?

File: spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/ReferenceLookupDelegate.java [hunk 4]
```diff
@@ -116,7 +116,7 @@ public ReferenceLookupDelegate(
 		}
 
 		if (!result.iterator().hasNext()) {
-			return null;
+			return property.isMap() ? Collections.emptyMap() : null;
```
Comment (schauder) [inline spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/convert/ReferenceLookupDelegate.java]: This makes me wonder if we have the same problem with collections like `Set` and `List`. 
Care to add even more tests to double check?

**Review by schauder — CHANGES_REQUESTED** [review]
I left a couple of comments.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java (hunk 2 above)
Comment (bamapookie) [reply spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java]: Done.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java (hunk 3 above)
Comment (bamapookie) [reply spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/MappingMongoConverterTests.java]: I wanted to ensure that the first `assertThat` call didn't have a side effect (such as lazy loading the map).  I can break this up into 2 tests.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/ReferenceLookupDelegateUnitTests.java [hunk 5]
```diff
@@ -94,4 +97,35 @@ void shouldResolveEmptyMapOnEmptyTargetCollection() {
 		lookupDelegate.readReference(property, Collections.emptyMap(), lookupFunction, entityReader);
 		verify(lookupFunction, never()).apply(any(), any());
 	}
+
+	@Test // GH-5065
+	@DisplayName("GH-5065: Empty Map with @DocumentReference annotation should deserialize to an empty map.")
```
[the remark this answers is not in the document]
Comment (schauder) [reply spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/ReferenceLookupDelegateUnitTests.java]: Let's remove them.

File: spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/DefaultDbRefResolverUnitTests.java [hunk 6]
```diff
@@ -134,4 +143,75 @@ void bulkFetchContainsDuplicates() {
...
+		when(property.getDocumentReference()).thenReturn(documentReference);
+		DocumentReferenceSource source = mock(DocumentReferenceSource.class);
+		when(source.getTargetSource()).thenReturn(Document.parse("{}"));
+		ReferenceLookupDelegate lookupDelegate = new ReferenceLookupDelegate(mappingContext, spELContext);
+
+		ReferenceResolver.MongoEntityReader entityReader = mock(ReferenceResolver.MongoEntityReader.class);
+
+		Object target = resolver.resolveReference(property, source, lookupDelegate, entityReader);
+
+		verify(property, times(3)).isMap();
```
[the remark this answers is not in the document]
Comment (schauder) [reply spring-data-mongodb/src/test/java/org/springframework/data/mongodb/core/convert/DefaultDbRefResolverUnitTests.java]: Let's remove them
